hamilton/cli/logic.py

```diff
@@ -0,0 +1,232 @@
+import functools
+import hashlib
+import subprocess
+from pathlib import Path
+from types import ModuleType
+from typing import Dict, List, Union
+
+from hamilton import ad_hoc_utils, driver, graph, graph_types, graph_utils
+
+
+def get_git_base_directory() -> str:
+    """Get the base path of the current git directory"""
+    try:
+        result = subprocess.run(
+            ["git", "rev-parse", "--show-toplevel"],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            text=True,
+        )
+
+        if result.returncode == 0:
+            return result.stdout.strip()
+        else:
+            print("Error:", result.stderr.strip())
+    except FileNotFoundError:
+        raise FileNotFoundError("Git command not found. Please make sure Git is installed.")
+
+
+def get_git_reference(git_relative_path: Union[str, Path], git_reference: str) -> str:
+    """Get the source code from the specified file and git reference"""
+    try:
+        result = subprocess.run(
+            ["git", "show", f"{git_reference}:{git_relative_path}"],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            text=True,
+        )
+
+        if result.returncode == 0:
+            return result.stdout.strip()
+        elif result.returncode == 128:
+            # TODO catch the following error:
+            # fatal: Path '{git_relative_path}' exists on disk, but not in '{git_revision}'.
+            return
+        else:
+            return
+    except FileNotFoundError:
+        raise FileNotFoundError("Git command not found. Please make sure Git is installed.")
+
+
+def version_hamilton_functions(module: ModuleType) -> Dict[str, str]:
+    """Hash the source code of Hamilton functions from a module"""
+    origins_version: Dict[str, str] = dict()
+
+    for origin_name, _ in graph_utils.find_functions(module):
+        origin_callable = getattr(module, origin_name)
+        origins_version[origin_name] = graph_utils.hash_source_code(origin_callable, strip=True)
+
+    return origins_version
+
+
+def hash_hamilton_nodes(dr: driver.Driver) -> Dict[str, str]:
+    """Hash the source code of Hamilton functions from nodes in a Driver"""
+    graph = graph_types.HamiltonGraph.from_graph(dr.graph)
+
+    nodes_version = dict()
+    for n in graph.nodes:
```

Comment:
Is this order going to be deterministic? I think dicts are, but might as well sort it as hashing is sensitive to order.

Reply:
This function is responsible for creating one hash per node, which should be independent. 

Next, `hash_dataflow` sorts the hashes and creates a single string from the list and encodes it to bytes. Not the most elegant, but it should work.

Did I missing something?

Reply:
Ahh got it. I'm wondering if there's a good reason to split it out? I imagine you're using both in different places. Makes sense.